python/cudf/cudf/core/scalar.py

```diff
@@ -16,7 +17,50 @@
 )
 
 
-class Scalar(BinaryOperand):
+# Note that the metaclass below can easily be generalized for use with
+# other classes, if needed in the future. Simply replace the arguments
+# of the `__call__` method with `*args` and `**kwargs`. This will
+# result in additional overhead when constructing the cache key, as
+# unpacking *args and **kwargs is not cheap. See the discussion in GH
+# #11246 for more details.
+class CachedScalarInstanceMeta(type):
+    """
+    Metaclass for Scalar that caches `maxsize` instances.
+    """
+
+    def __new__(self, names, bases, attrs, maxsize=128):
+        self.__maxsize = maxsize
```

Comment:
Unless I'm deeply misunderstanding the goal here, this code isn't doing what you want. In `__new__`, `self` will be the metaclass `CachedScalarInstanceMeta`, so you're effectively resetting the maxsize attribute to `CachedScalarInstanceMeta` every time a new class is created with `CachedScalarInstanceMeta` as its metaclass. For example:

```
import random


class Meta(type):
    def __new__(self, names, bases, attrs):
        self.__val = random.randint(0, 100)
        return type.__new__(self, names, bases, attrs)


class A(metaclass=Meta):
    pass

print(Meta._Meta__val)

class B(metaclass=Meta):
    pass

print(Meta._Meta__val)

# This prints two different random numbers for me each time.
```

I recognize that this is a moot point at present since the metaclass is only used in one place (and is somewhat specialized for it anyway due to the signature of `__call__`), but it's worth getting this right in case we do choose to generalize it in the future as per your comment above (or removing the feature if we don't think it generalizes). With that in mind, is the max size intended to be a constant across all classes using the metaclass, or should each class be able to set a different maxsize? If it should be settable, how would a subclass be expected to actually change its maxsize? I don't know how you would pass arguments to the metaclass constructor when defining a class.

Reply:
You're totally right -- this should just go in `__init__`, not `__new__`. Fixed.

> how you would pass arguments to the metaclass constructor when defining a class.

It turns out you can just do this:


```python
class Meta(type):
    def __new__(self, names, bases, attrs, x):
        return type.__new__(self, names, bases, attrs)

    def __init__(self, names, bases, attrs, x):
        self.x = x

class A(metaclass=Meta, x=1):
    pass

class B(metaclass=Meta, x=2):
    pass

print(A.x)  # prints 1
print(B.x)  # prints 2
```

Reply:
I didn't even consider that, but it makes sense. Cool!